### ISSUE TITLE
Replacing individual row clipping with a single one

### ISSUE THREAD

**Issue by NikolayNIK** [NikolayNIK/material_table_view#3]
One of the most severe performance bottlenecks is clipping operation performed for every visible row. Ideally clipping should be performed for all horizontally scrollable space only once. Drawing horizontally scrollable cell widgets in parallel  widget hierarchy would solve this, but it is not the solution as it's imperative for each row to contain all its cell widgets to enable material interactions across all of them.

The proposed solution is a bit intrusive. The idea is to extend the `PaintingContext` to allow it to hold multiple `PaintingContext` instances:
 - regular one (specifically for rows and frozen cell widgets);
 - clipped one (specifically for scrolling cell widgets);
 - regular one for placeholders (specifically for placeholder rows and frozen placeholder cell widgets);
 - clipped one for placeholders (specifically for scrolling placeholder cell widgets).

The last two are for applying shading to all the placeholders at once rather than for every row separately. This is now required and will lead to the need either:
 - to create extra clipped layer regardless of presence of visible placeholders (clipping operation might get optimized away either by Flutter or by Skia as it wouldn't contain any draw calls, this needs to be tested out);
 - or to create clipped layer lazily.

This way children will have an option to choose which `PaintingContext` to paint on, allowing the render object of a row to paint cell widgets on clipped either clipped or regular layer depending on the columns position.

I don't believe, Flutter is meant to be used that way, so I don't expect it to be particularly robust and consequence-free. One drawback I sese is that the described approach will limit the use of `PaintingContext`-changing widgets, such as `RepaintBoundary`, `ShaderMask`, various clipping ones and more in `rowBuilder` closure because the original will have to reach the row widget. While others can be implemented either by overloading `PaintingContext` layer creation methods to work with all layers at the same time or by providing alternative widgets specifically designed to work with that `PaintingContext` (that will limit the use of third party libraries in that context), I see no obvious solution to `RepaintBoundary`. The only use case I see for `PaintingContext` change for a row is placeholder shading. Directly exposing shading the layers (either only placeholder ones or both pairs) to the developer will cover those use cases. Also, providing out-of-the-box shading is an option.

In conclusion, this is a pretty restrictive and intrusive solution, but it will greatly benefit the intended use case and it might be the only solution.
